apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/DatastoreDbBuilder.cs

```diff
@@ -0,0 +1,247 @@
+        }
+
+        /// <summary>
+        /// The namespace ID, or null to use the default.
+        /// </summary>
+        public string NamespaceId { get; set; }
+
+        /// <inheritdoc />
+        public override DatastoreDb Build() =>
+            PrepareBuilder().Build();
+
+        /// <inheritdoc />
+        public override Task<DatastoreDb> BuildAsync(CancellationToken cancellationToken = default) =>
+            PrepareBuilder().BuildAsync(cancellationToken);
+
+        /// <inheritdoc />
+        protected override DatastoreDb BuildImpl() =>
+            throw new InvalidOperationException("Never called: Build is overridden");
+
+        /// <inheritdoc />
+        protected override Task<DatastoreDb> BuildImplAsync(CancellationToken cancellationToken) =>
+            throw new InvalidOperationException("Never called: BuildAsync is overridden");
+
+        /// <inheritdoc />
+        protected override ServiceEndpoint GetDefaultEndpoint() => DatastoreClient.DefaultEndpoint;
+
+        /// <inheritdoc />
+        protected override IReadOnlyList<string> GetDefaultScopes() => DatastoreClient.DefaultScopes;
+
+        private DatastoreClient CreateClient() => DatastoreClient.Create(CreateCallInvoker(), Settings);
+
+        private async Task<DatastoreClient> CreateClientAsync(CancellationToken cancellationToken) =>
+            DatastoreClient.Create(await CreateCallInvokerAsync(cancellationToken).ConfigureAwait(false), Settings);
+
+        private class EmulatorConfiguration
+        {
+            internal const string EmulatorHostVariable = "DATASTORE_EMULATOR_HOST";
+            internal const string EmulatorProjectVariable = "DATASTORE_PROJECT_ID";
+            internal ServiceEndpoint Endpoint { get; }
+            internal string ProjectId { get; }
+
+            private EmulatorConfiguration(ServiceEndpoint endpoint, string projectId)
+            {
+                Endpoint = endpoint;
+                ProjectId = projectId;
+            }
+
+            internal static EmulatorConfiguration FromEnvironment()
+            {
+                // Treat present-but-empty environment variables as if they were absent.
+                string hostAndPort = Environment.GetEnvironmentVariable(EmulatorHostVariable)?.Trim();
+                string projectId = Environment.GetEnvironmentVariable(EmulatorProjectVariable)?.Trim();
+                if (hostAndPort == "")
+                {
+                    hostAndPort= null;
+                }
+                if (projectId == "")
+                {
+                    projectId = null;
+                }
+                var endpoint = string.IsNullOrEmpty(hostAndPort) ? null : ParseEndpoint(hostAndPort);
+                return new EmulatorConfiguration(endpoint, projectId);
+            }
+
+            // TODO: Move this to ServiceEndpoint, and add lots of tests.
+            private static ServiceEndpoint ParseEndpoint(string text)
+            {
```

Comment:
Maybe try using System.Uri or System.UriBuilder instead of parsing. I'm almost certain it should work.